static/js/tools/timeline/data_fetcher.ts

```diff
@@ -221,77 +249,170 @@ export function fetchStatData(
+      denomData,
     };
-    const allDates = new Set<string>();
-    const statResp = allResp[0] as StatApiResponse;
-    const denomResp = allResp[1] as StatApiResponse;
-    const displayNameMapping = allResp[2] as DisplayNameApiResponse;
+  });
+}
+
+/**
+ * Given raw data and timeline chart options, gets the StatData for the chart.
+ * @param rawData raw timeline data
+ * @param places places to get data for
+ * @param statVars stat vars to get data for
+ * @param metahashMap map of stat var dcid to metahash
+ * @param isRatio whhether or not to calculate ratio
+ * @param denom denominator to use when calculating ratio
+ * @param scaling scaling factor to use when calculating ratio
+ */
+export function getStatData(
+  rawData: TimelineRawData,
+  places: string[],
+  statVars: string[],
+  metahashMap: Record<string, string>,
+  isRatio: boolean,
+  denom = "",
+  scaling = 1
+): StatData {
+  const result: StatData = {
+    places: places,
+    statVars: statVars,
+    dates: [],
+    data: {},
+    sources: new Set(),
+    measurementMethods: new Set(),
+  };
+
+  const allDates = new Set<string>();
 
-    for (const place in statResp) {
-      const placeData = statResp[place].data;
-      for (const statVar in placeData) {
-        if (isRatio) {
-          placeData[statVar] = computeRatio(
-            placeData[statVar],
-            denomResp[place].data[denom],
-            scaling
-          );
+  for (const place of places) {
+    const placeData = {};
+    for (const sv of statVars) {
+      let timeSeries = null;
+      const metahash = metahashMap[sv];
+      if (!_.isEmpty(metahash)) {
+        const metadata = rawData.metadataMap[sv][metahash];
+        const haveSeries =
+          place in rawData.statAllData &&
+          sv in rawData.statAllData[place] &&
+          metahash in rawData.statAllData[place][sv];
+        if (haveSeries) {
+          timeSeries = {
+            val: rawData.statAllData[place][sv][metahash].val,
+            metadata,
+          };
         }
+      } else if (
+        !_.isEmpty(rawData.statData[place]) &&
+        !_.isEmpty(rawData.statData[place].data[sv])
+      ) {
+        timeSeries = rawData.statData[place].data[sv];
       }
 
-      if (!placeData) {
+      if (_.isEmpty(timeSeries)) {
         continue;
       }
-      if (displayNameMapping[place]) {
-        statResp[place].name = displayNameMapping[place];
-      }
-      for (const statVar in placeData) {
-        const timeSeries = placeData[statVar];
-        if (timeSeries.metadata) {
-          result.sources.add(timeSeries.metadata.provenanceUrl);
-          const mmethod = timeSeries.metadata.measurementMethod;
-          if (mmethod) {
-            result.measurementMethods.add(mmethod);
-          }
+
+      if (isRatio) {
+        const hasDenomSeries =
+          place in rawData.denomData && denom in rawData.denomData[place].data;
+        if (!hasDenomSeries) {
+          continue;
```

Comment:
we should show an error message in this case. add a todo, or a simple message in below the chart?

Reply:
added a TODO. We should think more about what we want in the error message and when to show the error message. For example, if there are multiple places chosen and only one of the places is missing a denomSeries, I don't think we necessarily want to show an error message

Reply:
added a TODO. We should think more about what we want in the error message and when to show the error message. For example, if there are multiple places chosen and only one of the places is missing a denomSeries, I don't think we necessarily want to show an error message